lib/r10k/tarball.rb

```diff
@@ -0,0 +1,257 @@
+        tempfile.binmode
+        src_uri = URI.parse(source)
+
+        temp_digest = case src_uri.scheme
+                      when 'file', nil
+                        copy_to_file(src_uri.path, tempfile)
+                      when %r{[a-z]} # Windows drive letter
+                        copy_to_file(src_uri.to_s, tempfile)
+                      when %r{https?}
+                        download_to_file(src_uri, tempfile)
+                      else
+                        raise "Unexpected source scheme #{src_uri.scheme}"
+                      end
+
+        # Verify the download
+        unless (sha256digest == temp_digest) || sha256digest.nil?
+          raise 'Downloaded file does not match digest'
+        end
+
+        # Move the download to cache_path
+        FileUtils::mkdir_p(cache_root)
+        begin
+          FileUtils.mv(tempfile.path, cache_path)
+        rescue Errno::EACCES
+          FileUtils.cp(tempfile.path, cache_path)
+        end
+
+        tempfile.close
+        tempfile.unlink
+      end
+    end
+
+    # Is the cached tarball valid? Checks the cached tarball's digest against
+    # the expected digest.
+    #
+    # @return [Boolean]
+    def cache_valid?
+      unless sha256digest.nil?
+        File.exist?(cache_path) && sha256digest == file_digest(cache_path)
+      else
+        File.exist?(cache_path)
+      end
+    end
+
+    # List all of the files contained in the tarball and their paths. This is
+    # useful for implementing R10K::Purgable
+    #
+    # @return [Array] A list of file paths contained in the archive
+    def paths
+      names = Array.new
+      each_tarball_entry { |entry| names << entry.full_name.chomp('/') }
+      names
+    end
+
+    def cache_sha256digest
+      raise "Cache not present" unless File.exist?(cache_path)
+      file_digest(cache_path)
+    end
+
+    private
+
+    CHUNK_SIZE = 64 * 1024 # 64 kb
+
+    def each_tarball_entry(&block)
+      File.open(cache_path, 'rb') do |file|
+        Zlib::GzipReader.wrap(file) do |reader|
+          Archive::Tar::Minitar::Input.each_entry(reader) do |entry|
+            yield entry
+          end
+        end
+      end
+    end
+
+    # Return the sha256 digest of the file at the given path
+    #
+    # @param path [String] The path to the file
+    # @return [String] The file's sha256 hex digest
+    def file_digest(path)
+      File.open(path) do |file|
+        reader_digest(file)
+      end
+    end
+
+    # Return the sha256 digest of the readable data
+    #
+    # @param reader [String] An object that responds to #read
+    # @return [String] The read data's sha256 hex digest
+    def reader_digest(reader)
+      digest = Digest::SHA256.new
+      while chunk = reader.read(CHUNK_SIZE)
+        digest.update(chunk)
+      end
+
+      digest.hexdigest
+    end
+
+    def sanitized_dirname(string)
+      string.gsub(/[^@\w\.-]/, '-')
+    end
+
+    def http_get(uri, method: Net::HTTP::Get, redirect_limit: 10, &block)
```

Comment:
If you want this to be a generic http request method, I would change the method name to reflect that. Otherwise, if you want to keep it specific to GET, I would remove method as a param.

This also seems like something that should be shared, rather than specific to this class.

Reply:
Removed `method:` as a user opt, and hard-coded to `Net::HTTP::Get`.

Regarding sharing, at the moment I don't know of anything else in R10k that needs to download files this way. I've added a comment to the method which clarifies exactly what it is doing:

```
# Start a Net::HTTP::Get connection, then yield the Net::HTTPSuccess object
# to the caller's block. Follow redirects if Net::HTTPRedirection responses
# are encountered, and honor settings[:proxy].
```

I think this could potentially be generalized for sharing, but I'm hesitant to break it off unless/until there's another consumer in r10k that wants to use it, for fear of over engineering. As near as I can tell R10K currently offloads all download-esque activity to other libraries. I'm considering `Net::HTTP` to be such a library, though, it's probably possible to use Faraday instead... I think we already have that as a dependency, if it gets us something `Net::HTTP` doesn't.

Reply:
Reviewing Faraday, it looks like that library can do everything we need it to and more succinctly than `Net::HTTP`. Given that it's already an r10k dependency, the right thing to do may be to reduce this method to using Faraday instead.

Reply:
I took a stab at using Faraday, but it looks like:

1. We want/need [streaming downloads](https://lostisland.github.io/faraday/usage/streaming), so I think we need a newer version of Faraday than r10k currently supports, due to the puppet-forge gem dependency constraints.
2. Using the newer version of Faraday, we run into a bit of a problem for testing; I can't figure out how to make Faraday's test stubs work with streaming downloads.

If these two problems didn't exist, it seems like we could fairly handily offload this to Faraday. I'm stopping that line of investigation for now though. Seems pointless given item 1.

If it did work, it _might_ look something like this.

```ruby
# @param uri [URI] The URI to download from
# @param output The file or path to save to
# @return [String] The downloaded file's sha256 hex digest
def download_to_file(uri, output)
  digest = DIGEST_CLASS.new

  File.open(output, 'wb') do |output_stream|
    faraday_connection.get(uri) do |req|
      req.options.on_data = Proc.new do |chunk, _|
        output_stream.write(chunk)
        digest.update(chunk)
      end
    end
  end

  digest.hexdigest
end

# Defined separately to enable test stubbing
def faraday_connection
  Faraday.new do |faraday|
    faraday.response(:follow_redirects)
    faraday.adapter(:net_http)
    faraday.proxy(settings[:proxy]) unless settings[:proxy].nil?
  end
end
```

Reply:
Alright, last tweak for now: went ahead and moved all the file download/copy/checksum related methods out of `R10K::Tarball` and into a new utility module, `R10K::Util::Downloader`.